src/codegen/compile.ml

```diff
@@ -8475,9 +8479,25 @@ module FuncDec = struct
       | Type.Shared Type.Query ->
         Lifecycle.trans env Lifecycle.PostQuery
       | Type.Shared Type.Composite ->
-        Lifecycle.trans env Lifecycle.Idle
+        (* Stay in composite query state such that callbacks of 
+        composite queries can also use the memory reserve. 
+        The state is isolated since memory changes of queries 
+        are rolled back by the IC runtime system. *)
+        Lifecycle.trans env Lifecycle.InComposite
       | _ -> assert false
 
+  let callback_start env =
+    Lifecycle.is_in env Lifecycle.InComposite ^^
+    G.if0
+      (G.nop)
+      (message_start env (Type.Shared Type.Write))
+
+  let callback_cleanup env =
+    Lifecycle.is_in env Lifecycle.InComposite ^^
+    G.if0
+      (G.nop)
```

Comment:
So I guess this means we never do GC in a composite query, since the GC increment is only invoked by message_cleanup env (Type.Shared Type.Write). 
I'm not sure - do we want that behaviour or not? I can see the argument for not doing the GC at return from a composite query. But this means we never do any GC during a composite query, I think. It's probably fine, but just wanted to point it out.

Reply:
Re-reading the PR description it look like this is deliberate.  Fine by me, I think.

Reply:
This a good point. I also thought about this and I do not have a clear answer. 

On the one hand, GC could make in composite query if they are very garbage-intense and have a lot of callbacks which I am not sure whether this is a typical case in practice. 

On the other hand, the state changes are discarded anyway at the end of composite query and the costs of the temporary garbage collection may be unjustified. If I remember right, the state is even reverted between repeated composite query calls to the same canister. 

Because of this short-lived nature, I had a slight preference for skipping the GC in favor of cheaper composite queries...